tensorzero-core/src/optimization/gcp_vertex_gemini_sft/mod.rs

```
@@ -0,0 +1,460 @@
+    pub n_epochs: Option<usize>,
+    pub export_last_checkpoint_only: Option<bool>,
+    #[cfg_attr(test, ts(type = "string | null"))]
+    pub credentials: Option<CredentialLocation>,
+    pub api_base: Option<Url>,
+    pub seed: Option<u64>,
+    pub service_account: Option<String>,
+    pub kms_key_name: Option<String>,
+    pub tuned_model_display_name: Option<String>,
+    pub bucket_path_prefix: Option<String>, // e.g., "fine-tuning/data" (optional)
+}
+
+impl std::fmt::Display for UninitializedGCPVertexGeminiSFTConfig {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        let json = serde_json::to_string_pretty(self).map_err(|_| std::fmt::Error)?;
+        write!(f, "{json}")
+    }
+}
+
+#[cfg(feature = "pyo3")]
+#[pymethods]
+impl UninitializedGCPVertexGeminiSFTConfig {
+    #[expect(clippy::too_many_arguments)]
+    #[new]
+    #[pyo3(signature = (*, model, bucket_name, project_id, region, learning_rate_multiplier=None, adapter_size=None, n_epochs=None, export_last_checkpoint_only=None, credentials=None, api_base=None, seed=None, service_account=None, kms_key_name=None, tuned_model_display_name=None, bucket_path_prefix=None))]
+    pub fn new(
+        model: String,
+        bucket_name: String,
+        project_id: String,
+        region: String,
+        learning_rate_multiplier: Option<f64>,
+        adapter_size: Option<usize>,
+        n_epochs: Option<usize>,
+        export_last_checkpoint_only: Option<bool>,
+        credentials: Option<String>,
+        api_base: Option<String>,
+        seed: Option<u64>,
+        service_account: Option<String>,
+        kms_key_name: Option<String>,
+        tuned_model_display_name: Option<String>,
+        bucket_path_prefix: Option<String>,
+    ) -> PyResult<Self> {
+        // Use Deserialize to convert the string to a CredentialLocation
+        let credentials =
+            credentials.map(|s| serde_json::from_str(&s).unwrap_or(CredentialLocation::Env(s)));
```

This should propagate the error with `?`, rather than using `unwrap_or`